hibernate-core/src/main/java/org/hibernate/query/sqm/function/NamedSqmFunctionDescriptor.java

```diff
@@ -6,59 +6,91 @@
+import org.hibernate.sql.ast.spi.SqlAppender;
+import org.hibernate.sql.ast.tree.SqlAstNode;
 
 /**
- * Provides a standard implementation that supports the majority of the HQL
- * functions that are translated to SQL. The Dialect and its sub-classes use
- * this class to provide details required for processing of the associated
- * function.
+ * A {@link SqmFunctionDescriptor function descriptor} for functions
+ * which produce SQL in the standard form {@code f(x, y, z)}.
  *
  * @author David Channon
  * @author Steve Ebersole
  */
-public class NamedSqmFunctionDescriptor extends AbstractSqmFunctionDescriptor {
+public class NamedSqmFunctionDescriptor
+		extends AbstractSqmSelfRenderingFunctionDescriptor
+		implements FunctionRenderingSupport {
 	private final String functionName;
-	private final FunctionRenderingSupport renderingSupport;
+	private String sqlFunctionName;
+	private final boolean requiresArguments;
+	private final String argumentListSignature;
 
 	public NamedSqmFunctionDescriptor(
 			String functionName,
-			boolean useParenthesesWhenNoArgs,
-			ArgumentsValidator argumentsValidator) {
-		this(
-				functionName,
-				useParenthesesWhenNoArgs,
-				argumentsValidator,
-				(impliedTypeAccess, arguments) -> impliedTypeAccess.get()
-		);
-	}
-
-	public NamedSqmFunctionDescriptor(
-			String functionName,
-			boolean useParenthesesWhenNoArgs,
+			String sqlFunctionName,
+			boolean requiresArguments,
 			ArgumentsValidator argumentsValidator,
-			FunctionReturnTypeResolver returnTypeResolver) {
-		super( argumentsValidator, returnTypeResolver );
+			FunctionReturnTypeResolver returnTypeResolver,
+			String argumentListSignature) {
+		super( functionName, argumentsValidator, returnTypeResolver );
 
 		this.functionName = functionName;
-		this.renderingSupport = new StandardFunctionRenderingSupport( useParenthesesWhenNoArgs );
+		this.sqlFunctionName = sqlFunctionName;
+		this.requiresArguments = requiresArguments;
+		this.argumentListSignature = argumentListSignature;
 	}
 
 	public String getFunctionName() {
 		return functionName;
 	}
 
 	@Override
-	protected String resolveFunctionName(String functionName) {
-		return this.functionName;
+	public FunctionRenderingSupport getRenderingSupport() {
+		return this;
 	}
 
 	@Override
-	public FunctionRenderingSupport getRenderingSupport() {
-		return renderingSupport;
+	public void render(
+			SqlAppender sqlAppender,
+			String functionName,
+			List<SqlAstNode> sqlAstArguments,
+			SqlAstWalker walker,
+			SessionFactoryImplementor sessionFactory) {
+		final boolean useParens = requiresArguments || !sqlAstArguments.isEmpty();
+
+		sqlAppender.appendSql( sqlFunctionName );
+		if ( useParens ) {
+			sqlAppender.appendSql( "(" );
+		}
+
+		boolean firstPass = true;
+		for ( SqlAstNode sqlAstArgument : sqlAstArguments ) {
+			if ( !firstPass ) {
+				sqlAppender.appendSql( ", " );
+			}
+			sqlAstArgument.accept( walker );
+			firstPass = false;
```

Comment:
We could move the above statement into the `if` block to save some unnecessary CPU cycles.